MODFLOW-USGS/modflow6
# refactor: tests

This PR builds on work [here](https://github.com/mjreno/modflow6/commit/be8f21aca780f1d5f1bf6b8f7cf0ec02ac251b52) to update tests and integrate `modflow-devtools`. Changes include:

* add `pytest.ini` with markers
* add `--smoke` test flag (`-S`)
* refactor `targets.py` as a fixture
* refactor `common_regression` with directory-scoped `conftest.py`
* use keepable temp dir fixtures instead of manually managed `temp` folder
* rename `Simulation` to `TestConfig` to disambiguate from FloPy's `Simulation`
* replace `pymake` in tests with `modflow-devtools` (`pymake` still required for other utility scripts)
* remove intel copyright notices from CI workflows (moved to [compiler install action](https://github.com/modflowpy/install-intelfortran-action))
* remove teardown functionality from testing framework (handled by `pytest`)
* remove `build_mf6_models_legacy` from testing framework (no longer needed)
* make `flopy` imports mandatory (`flopy` is a core testing dependency)
* use `pytest_generate_tests` to generate cases for `z0*` tests
* incremental refactor with `pytest-cases`
* reorganize CI workflows (add smoke test step, colocate lint/build/smoke/test jobs)
* bump `actions/cache` to `v3` in CI workflows
* update `DEVELOPER.md`

## Framework refactor

### Tests defined in code

Most of the test cases currently construct one or more models in code, in the same test script as the test function. The test function is for the most part identical between test scripts, inviting de-duplication. 

A subset of the multi-aquifer well tests (`test_gwf_maw[01-03].py`) are refactored with `pytest-cases`. The test function is [factored out](https://github.com/MODFLOW-USGS/modflow6/blob/16cbfb9b6c81562d56e09b93e8056267b5cafc30/autotest/test_gwf.py) and shared by a collection of [reusable test cases](https://github.com/MODFLOW-USGS/modflow6/blob/16cbfb9b6c81562d56e09b93e8056267b5cafc30/autotest/cases/gwf/maw.py). Cases are injected as 4-tuples including:

- case data
- the main simulation
- a comparison simulation (optional)
- an evaluation function (optional)

Though the test function can be shared by an arbitrary number of test cases, cases can still be selected by name and run independently with pytest's `-k` option.

A few other tests are updated to use `pytest-cases`, without having been moved from their original locations:

- [`test_gwf_maw04.py`](https://github.com/MODFLOW-USGS/modflow6/blob/16cbfb9b6c81562d56e09b93e8056267b5cafc30/autotest/test_gwf_maw04.py)
- [`test_gwf_buy_lak02.py`](https://github.com/MODFLOW-USGS/modflow6/blob/16cbfb9b6c81562d56e09b93e8056267b5cafc30/autotest/test_gwf_buy_lak02.py)
- [`test_gwf_csub_sk01.py`](https://github.com/MODFLOW-USGS/modflow6/blob/16cbfb9b6c81562d56e09b93e8056267b5cafc30/autotest/test_gwf_csub_sk01.py)

These can be plugged into a shared test function as above or can run independently (`maw04` is plugged into `test_gwf.py`, the latter two are still standalone). This is meant to explore how a refactor could proceed incrementally, keeping restructuring minimal, changes manageable, and tests passing. The framework can probably be simplified further with a few more iterations and with the help of the test context object introduced in `modflow-devtools`.

### Example models and test models

The `z0*` tests are moved to their own directory to allow a nested `conftest.py`. I wasn't sure what to call this folder &mdash; these tests differ from most of the others in loading test models from external repositories instead of creating them programmatically (the example models are an intermediate case, since they need to be built from the repo data first). Maybe `examples` works as an umbrella term? For now I just called it `autotest/z0/`.

Previously test cases were built independently in each test script. Now cases are loaded in the `pytest_generate_tests` hook in `autotest/z0/conftest.py` and can be requested from test functions by adding parameters `example`, `testmodel` or `largetest`.

Previously the test scripts would walk the filesystem looking for the example/largetest/testmodel repositories, adding several seconds of overhead during test collection. Now the tests expect these repos to live side-by-side with (in the same parent directory as) the `modflow6` repo &mdash; if not found there, the tests are skipped. This seems like a reasonable convention as it is straightforward to [checkout repos side-by-side in CI](https://github.com/actions/checkout#Checkout-multiple-repos-side-by-side) with `actions/checkout` (example [here](https://github.com/MODFLOW-USGS/modflow6/blob/16cbfb9b6c81562d56e09b93e8056267b5cafc30/.github/workflows/examples.yml)). The pattern I tried in `flopy`, where the test framework would automatically download example models to a temporary location, has also [proven a bit fragile](https://github.com/modflowpy/flopy/pull/1554#issue-1380325842) (bottom comment). Despite protection with a file lock, the download is occasionally corrupted, causing errors during collection (this likely happens if the user kills a test session while the download is in progress). Downloading fresh for every session incurs a lot of overhead, and adding code to handle errors and re-download seems less straightforward than simply documenting that developers need to clone the repos next to `modflow6` or `flopy` before running the tests. If this seems reasonable I can update the `flopy` tests accordingly.

## Pytest CLI options

A `--smoke` flag (short `-S`) is added to run a fast subset of the tests. Smoke tests are $\overline { ( slow \cup regression ) } $.

The `common_regression` functionality to select models by name and packages used is now supported with pytest options `--model` and `--package`. These options are only relevant for `z0*` tests.

### Filtering by model name

Collecting large model tests with no filters:

```shell
autotest % pytest -v z0/test_largetestmodels.py --collect-only
...
collected 18 items                                                                                                                                                                                                
```

To select a particular model:

```shell
autotest % pytest -v z0/test_largetestmodels.py --collect-only --model test1002_biscqtg_disv_gnc_nr_dev
...
collected 1 item                                                                                                                                                                                                  

<Package z0>
  <Module test_largetestmodels.py>
    <Function test_large_model[test1002_biscqtg_disv_gnc_nr_dev]>
```

The same can also be accomplished with pytest's `-k` option &mdash; however, the `--model` option can be used multiple times, e.g. `--model <model 1> --model <model 2>`.

### Filtering by package

Test cases can also be filtered by packages used. For instance, to select only test cases with a multi-aquifer well package:

```shell
autotest % pytest -v z0/test_largetestmodels.py --collect-only --package maw
...
collected 1 item                                                                                                                                                                                                  

<Package z0>
  <Module test_largetestmodels.py>
    <Function test_large_model[test1003_MNW2_Fig28]>
```
